lib/Dialect/FIRRTL/Transforms/GrandCentralSignalMappings.cpp

```diff
@@ -0,0 +1,322 @@
+  StringRef localName;
+  StringRef localFields;
+
+  /// The type of the signal being mapped.
+  FIRRTLType type;
+  /// The `localName` and `localFields` resolved to an actual value.
+  Value localValue;
+};
+
+struct ModuleSignalMappings {
+  ModuleSignalMappings(FModuleOp module) : module(module) {}
+  void run();
+  void addTargets(ArrayAttr targets, MappingDirection dir);
+  void emitMappingsModule();
+  void instantiateMappingsModule();
+
+  FModuleOp module;
+  bool anyFailed = false;
+  SmallVector<SignalMapping> mappings;
+  SmallString<64> mappingsModuleName;
+
+  /// A lookup table of local operations that carry a name which can be
+  /// referenced in an annotation.
+  DenseMap<StringRef, Value> localNames;
+};
+} // namespace
+
+template <typename T>
+static T &operator<<(T &os, const SignalMapping &mapping) {
+  os << "SignalMapping { remote"
+     << (mapping.dir == MappingDirection::DriveRemote ? "Sink" : "Source")
+     << ": " << mapping.remoteTarget << ", "
+     << "localTarget: \"" << mapping.localName << "\" }";
+  return os;
+}
+
+void ModuleSignalMappings::run() {
+  LLVM_DEBUG(llvm::dbgs() << "Running on module `" << module.getName()
+                          << "`\n");
+
+  // Gather the signal driver annotations on this module, and keep track of the
+  // target within the module.
+  LLVM_DEBUG(llvm::dbgs() << "- Gather annotations\n");
+  AnnotationSet::removeAnnotations(module, [&](Annotation anno) {
+    if (!anno.isClass(signalDriverAnnoClass))
+      return false;
+    if (auto sinks = anno.getMember<ArrayAttr>("sinkTargets"))
+      addTargets(sinks, MappingDirection::DriveRemote);
+    if (auto sources = anno.getMember<ArrayAttr>("sourceTargets"))
+      addTargets(sources, MappingDirection::ProbeRemote);
+    return true;
+  });
+  if (anyFailed)
+    return;
+
+  // Nothing to do if there are no signal mappings.
+  if (mappings.empty()) {
+    LLVM_DEBUG(llvm::dbgs() << "- No annotations, nothing left to do\n");
+    return;
+  }
+
+  // Gather a name table to resolve local mappings.
+  module.walk([&](Operation *op) {
+    TypeSwitch<Operation *>(op).Case<WireOp, NodeOp, RegOp, RegResetOp>(
+        [&](auto op) {
+          localNames.insert({op.name(), op});
+        });
+  });
+
+  // Gather the types and values for the signal mappings. This may incur
+  // additional subfield/subfindex operations to be generated.
+  ImplicitLocOpBuilder builder(module.getLoc(), module);
+  builder.setInsertionPointToEnd(module.getBodyBlock());
+  for (auto &mapping : mappings) {
+    // Resolve the name locally.
+    mapping.localValue = localNames.lookup(mapping.localName);
+    if (!mapping.localValue) {
+      module.emitError("unknown local target \"")
+          << mapping.localName << "\" in SignalDriverAnnotation";
+      anyFailed = true;
+      return;
+    }
+    LLVM_DEBUG(llvm::dbgs() << "- Resolved `" << mapping.localName << "` to "
+                            << mapping.localValue << "\n");
```

Comment:
Yeah totally. I've implemented this way because it's not quite clear to me yet how this will interact with multi-circuit stuff and XMR operations. Scattering the annotations would require us to pick whether to scatter them to the source or destination target, or both, but after XMRs are in we might decide to replace the annotation with an op itself, or have the annotation include a symbol itself.

From a pass implementation point of view, I think it would actually be beneficial to *not* scatter the annotation (also applies to data taps), but actually have the circuit or module act as a "table" where the `SignalDriverAnnotation`s can easily be accessed. Otherwise we have to do this traversal of the IR, find the two halves that the anno was split and scattered to, and link them back up again.

Maybe having symbols in the annotation could be a thing? But that's probably hard for passes to then go in and replace symbol uses also in the annos. So this annotation could possibly live in an operation, where the symbol replacement could work?

Reply:
Actually you're right: the current implementation assumes the annotation to be scattered to the module. So I'm already making an assumption about how this stuff gets scattered. So might as well scatter it to the target operation properly.

Reply:
This is now properly scattered.